android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java

```diff
@@ -219,12 +225,14 @@ public synchronized void close() {
             if (Logging.WARNING) Log.w(Logging.TAG, "output close failure", e);
         }
         try {
-            mSocket.close();
+            if (mTcpSocket != null)  mTcpSocket.close();
```

Comment:
I'd put `mTcpSocket.close()` and `mSocket.close()` into different `try...catch` block because otherwise if first call (in this case `mTcpSocket.close()` fails, `mSocket.close()` won't be called).

Reply:
as the logic now, even in the Idiot-proof design, you cannot have both allocate mTcpSocket and mSocket even when you switch. While you connect to one of them (when you switch), you clean the system.

Reply:
I added.